Jetpack focus: Disable push notifications for WP

ravishanker opened the pull request:
If an instance of JP is detected then disable push notifications for WP to prevent double notifications

- [x] Trigger this from Jetpack app, on notifications screen in the welcome flow in a separate PR ( [See example](https://github.com/wordpress-mobile/WordPress-Android/pull/17400/commits/4b6c7caa12eea6d1ff8486fc43d36bd9fc5d4244) )
- [x] Add permissions to exported BroadcastReceiver.  Preferably, `signature`, otherwise `normal` in a separate PR

To test:

Test 1:
- Ensure either Jetpack app or Jetpack beta app is installed
- Ensure WordPress app is also on the device
- Ensure you already receive some push notifications

Test 2:
- Remove any WordPress app instances
- Build and Run WordPress app or install using QR code below
- Open Terminal in AS or Terminal app. Execute   `adb shell am broadcast -p org.wordpress.android.prealpha -a org.wordpress.android.broadcast.DISABLE_NOTIFICATIONS` command ([Welcome flow ](https://github.com/wordpress-mobile/WordPress-Android/pulls/ovitrif)is still in the works, till then use command line to test it on WP side)
- Verify `Action: org.wordpress.android.broadcast.DISABLE_NOTIFICATIONS` is displayed in a toast message on the app.  The same can be seen in the Logcat in AS
- Verify switch is set to Off on Notification Settings ( Me > Notifications tab > ⚙️ on the top right hand corner )
- Wait till the next day if required or [push a debug notification](https://mc.a8c.com/mobile/push-notifications/debug/)
- Verify push notifications are **not** displayed for WordPress app
- Ensure notifications are displayed for Jetpack app as before.


NOTE:  this doesn't cover any local notifications e.g. WeeklyRoundup


## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually

3. What automated tests I added (or what prevented me from doing so)
Existing tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
Disable push notifications for WP

If an instance of JP is detected then disable push notifications for WP to prevent double notifications

ravishanker committed:
Apply feature flag

apply feature flag jetpack_powered_bottom_sheet_remote_field

ravishanker committed:
Remove debug logging

ravishanker committed:
Remove unused import

staskus commented on WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:
```diff
@@ -459,6 +461,20 @@ && getIntent().getExtras().getBoolean(ARG_CONTINUE_JETPACK_CONNECT, false)) {
+        if (mJetpackBrandingUtils.shouldShowJetpackPoweredBottomSheet()
+            && !mBuildConfigWrapper.isJetpackApp()
+            && isJetpackInstalled()) {
+            mGCMMessageHandler.removeAllNotifications(this);
```
Once this disables notifications on `WordPress` app, does it get reflected in the `Notification Settings` view, showing disabled `On` flag?

ravishanker replied:
Good point.  this PR is only aimed at testing the disabling notifications actually works for now.

ravishanker committed:
Turn off Notifications Settings switch

Turn off switch on Notifications Settings screen

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:
```diff
@@ -459,6 +463,25 @@ && getIntent().getExtras().getBoolean(ARG_CONTINUE_JETPACK_CONNECT, false)) {
+        }
+    }
+
+    private boolean isJetpackInstalled() {
```
I suggest `isJetpackAppInstalled` to name the function since there is already an [isJetpackInstalled](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/model/SiteModel.java#L597) function which checks Jetpack plugin.

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:
```diff
@@ -459,6 +463,25 @@ && getIntent().getExtras().getBoolean(ARG_CONTINUE_JETPACK_CONNECT, false)) {
         }
         mSharedLoginResolver.tryJetpackLogin();
+
+        if (mJetpackBrandingUtils.shouldShowJetpackPoweredBottomSheet()
```
`mJetpackBrandingUtils.shouldShowJetpackPoweredBottomSheet()` has 3 different checks,
https://github.com/wordpress-mobile/WordPress-Android/blob/4cced645e5f360c30561333a0654d8659b70e732/WordPress/src/main/java/org/wordpress/android/util/JetpackBrandingUtils.kt#L23-L25
I think these are not appropriate for our case.
We talked about an independent remote flag for this project with @guarani, and we thought we might want to turn off this project anytime, it'd be better to have an independent remote flag for just disabling notifications.

irfano reviewed (changes requested):
Thanks, Ravi! This only removed received notifications when the WordPress app is launched. We will prevent receiving new WP notifications. But there is a case that users can install the Jetpack app when they already have received WP notifications. This PR handles only this case, right?

The function works on my test. 👍🏻  I just commented for some code changes.

ravishanker committed:
Add JetpackAppInstallReceiver

Updating implementation to use BroadcastReceiver to disable notifications

ravishanker committed:
Update WPMainActivity.java

Remove initial implementation in favour of BroadcastReceiver

ravishanker commented:
@irfano - I've updated the implementation to use a [BroadcastReceiver](2a1bac579ee37db22e93ef9d8fe2632d5c9ef1b4) to trigger this from Jetpack app, after install, login, and when in the welcome flow.  wdyt?

This approach has advantage of removing the need for detecting Jetpack app install, login, and inform steps.

[Welcome flow ](https://github.com/wordpress-mobile/WordPress-Android/pulls/ovitrif)is still in the works, till then we can hook this somewhere on the Jetpack app side temporarily to test it out.  Updated instructions in the description

ravishanker committed:
Update WPMainActivity.java

Remove unnecessary injected class

ravishanker committed:
Add intent filter to Receiver

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/jetpackbadge/JetpackAppInstallReceiver.kt:
```diff
@@ -0,0 +1,53 @@
+    @Inject lateinit var gcmMessageHandler: GCMMessageHandler
+
+    override fun onReceive(context: Context?, intent: Intent?) {
+        (context?.applicationContext as WordPress).component().inject(this)
```
We can remove this `inject` line and add `@AndroidEntryPoint` to the top of the class.

ravishanker replied:
Yes, Updated (Couldn't use constructor injection for now.  It wouldn't work with adb commands)

ravishanker replied:
> I couldn't receive push notifications for WP, even when the switch was on

It's working for me.  Choose Application(WordPress Android), Username (Yours), and Device (All Devices) if you're using [Debug Console](https://mc.a8c.com/mobile/push-notifications/debug/) to test.  Also, come out of Notifications settings screen if it helps!

irfano replied:
I changed the variant to `jalapeno` and it worked. 👍🏻

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/jetpackbadge/JetpackAppInstallReceiver.kt:
```diff
@@ -0,0 +1,53 @@
+            }
+        }
+
+        if (intent != null && !mBuildConfigWrapper.isJetpackApp) {
```
- Should we also check if WP app is logged in? 
- We can add the data of the logged-in account id of JP to broadcast intent and send it to WP. Then we can check if WP account is the same. If it's not the same, we don't disable notifications. Wdyt?

ravishanker replied:
Probably not, given this flow is from WP -> to JP and since origin is already in logged in.

ravishanker replied:
Build is failing due to lint as it needs a permission defined like so in WP 
    `<permission
        android:name="org.wordpress.android.permission.DISABLE_NOTIFICATIONS"
        android:label="@string/notifications_disabled"
        android:description="@string/notifications_disabled"
        android:protectionLevel="normal" />`

and on the `<receiver `
`android:permission="org.wordpress.android.permission.DISABLE_NOTIFICATIONS"`

**Not able to use adb command for testing after adding permission, so left it for now to finish testing**
`adb shell am broadcast -p org.wordpress.android.prealpha -a org.wordpress.android.broadcast.DISABLE_NOTIFICATIONS --receiver-permission org.wordpress.android.permission.DISABLE_NOTIFICATIONS`

Need to also add the following on the Jetpack app side while invoking
Add the following to Jetpack app AndroidManifest.xml
`<uses-permission android:name="org.wordpress.android.permission.DISABLE_NOTIFICATIONS"/>`

and then
`Intent().also { intent -> intent.action = "org.wordpress.android.broadcast.DISABLE_NOTIFICATIONS" context?.sendBroadcast(intent, "org.wordpress.android.permission.DISABLE_NOTIFICATIONS") }`

irfano reviewed (changes requested):
Looks great! This is a clever solution. 👏🏻 
I tested sending broadcast, and it worked. The notification switch was turned off successfully. But while testing, I couldn't receive push notifications for WP, even when the switch was on. I'll test it again.

ravishanker committed:
Use AndroidEntryPoint

staskus commented:
> I've updated the implementation to use a [BroadcastReceiver](https://github.com/wordpress-mobile/WordPress-Android/pull/2a1bac579ee37db22e93ef9d8fe2632d5c9ef1b4) to trigger this from Jetpack app, after install, login, and when in the welcome flow. wdyt?

Great news! iOS and Android solutions could pretty much match then. We can even align on conditions when to trigger this from Jetpack app 👍

ravishanker committed:
ignore lint check for now

ravishanker committed:
Update JetpackAppInstallReceiver.kt

removed TODO for detekt

irfano committed:
Remove redundant inject in AppComponent

irfano committed:
Move JetpackAppInstallReceiver to wordpress directory

This receiver will be used only in WordPress app.

irfano committed:
Add DISABLE_NOTIFICATIONS permission for broadcast

irfano committed:
Remove debug logging from JetpackAppInstallReceiver

irfano committed:
Update DISABLE_NOTIFICATIONS permission label and desc

irfano committed:
Merge branch 'trunk' into Jetpack-Focus-Disable-notifications-for-WP-on-JP-Install

# Conflicts:
#	WordPress/src/wordpress/AndroidManifest.xml

irfano commented:
- I moved `JetpackAppInstallReceiver` to `wordpress` directory. Since it will work only in WP app,
- I added `org.wordpress.android.permission.DISABLE_NOTIFICATIONS` permission,
- Removed debug logging,
- Synced with trunk,
- Tested it with this code block, and it worked. But the package name should align with the build variant.
```
Intent().also { intent ->
    intent.action = "org.wordpress.android.broadcast.DISABLE_NOTIFICATIONS"
    intent.package = "org.wordpress.android"   
    context?.sendBroadcast(intent, "org.wordpress.android.permission.DISABLE_NOTIFICATIONS")
}
```

ravishanker commented:
>     * Tested it with this code block, and it worked. But the package name should align with the build variant.

I tried something like this earlier
```
val appSuffix = BuildConfig.APPLICATION_ID.split(".").last()
            val appPackage = if (appSuffix.isNotBlank()) {
                "org.wordpress.android.${appSuffix}"
            } else {
                "org.wordpress.android"
            }
            intent.setPackage(appPackage)
            AppLog.d(T.NOTIFS, intent.toString())
```

irfano committed:
Merge branch 'trunk' into Jetpack-Focus-Disable-notifications-for-WP-on-JP-Install
